Kernel/Devices/USB/USBPipe.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2021 Jesse Buhagiar <jooster669@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/OwnPtr.h>
+#include <AK/Types.h>
+#include <Kernel/Devices/USB/USBDescriptors.h>
+#include <Kernel/VM/Region.h>
+
+namespace Kernel::USB {
+
+//
+// A pipe is the logical connection between a memory buffer on the PC (host) and
+// an endpoint on the device.
+// https://github.com/haiku/haiku/blob/master/src/add-ons/kernel/bus_managers/usb/usb_private.h
+//
+class USBPipe {
```

Comment:
I don't get the purpose of this class. You don't do anything meaningful with it. Is it just here because you saw that Haiku has an equivalent?

Reply:
[not](https://github.com/PaulStoffregen/antplus/blob/master/USBHost_t36.h#L153) [just](https://github.com/harlequin-tech/arduino-usb/blob/master/LUFA/Drivers/USB/LowLevel/Pipe.h) [haiku](https://www.jungo.com/st/support/documentation/windriver/10.2.1/wdusb_manual.mhtml/USB_data_flow.html) [does](https://docs.microsoft.com/en-us/windows-hardware/drivers/wdf/working-with-usb-pipes) this. How else do you suppose we structure this then? The spec states that:

`A USB pipe is an association between an endpoint on a device and software on the host. Pipes represent the
ability to move data between software on the host via a memory buffer and an endpoint on a device. There
are two mutually exclusive pipe communication modes`. When the stack is more fleshed out, an pipe will be held for each endpoint in each direction. In my mind this information shouldn't be contained anywhere else. I didn't just copy this from Haiku.

This class describes the method by which the transfer occurs, either toward or from the host. Perhaps the buffer should be held here instead of containing it within `USBTransfer`?

Reply:
How else do you suppose we structure this then? The spec states that:

A USB pipe is an association between an endpoint on a device and software on the host. Pipes represent the ability to move data between software on the host via a memory buffer and an endpoint on a device. There are two mutually exclusive pipe communication modes. When the stack is more fleshed out, a pipe will be held for each endpoint in the direction that the endpoint represents. In my mind this information shouldn't be contained anywhere else. I didn't just copy this from Haiku, though seeing how they structured theirs helped glue a little bit of how it all worked together in my head.

This class describes the method by which the transfer occurs, either toward or from the host, as well as the type of endpoint pipe it is (control, bulk, interrupt or Isochronous). Perhaps the buffer should be held here instead of containing it within USBTransfer (this probably makes more sense), and any of this functionality be moved into some USBEndpoint class? Considering what a pipe describes, however, this in my opinion feels wrong.

Have I misunderstood how this is supposed to work exactly (or misread the spec)?

Reply:
@awesomekling @linusg Considering this person has deleted their account, I'm going to ignore this comment and baseless claim of misunderstanding how the bus works and copying from other implementations; most I see seem to use the pipe opaquely in this manner and the fact that another implementation uses a class isn't a reason for such an accusation provided with literally no explanation. I feel I've provided enough justification in this comment. `USBPipe` here is used to "connect" the endpoint to the `USBTransfer` buffer, it's not "doing nothing meaningful".

If somebody else that knows more about USB than myself sees that this is an incorrect way of handling this as the design scales, then they can fix it or discuss the design on the Discord :)